src/compiler/parser.ts

```diff
@@ -6356,78 +6365,111 @@ namespace ts {
+                    scanChildTags(tryParseChildTag, callbackTypeNode);
+                    callbackTag.type = finishNode(callbackTypeNode);
+                    return finishNode(callbackTag);
+
+                    function tryParseChildTag(parentTag: JSDocCallbackType) {
+                        Debug.assert(token === SyntaxKind.AtToken);
+                        const atToken = createNode(SyntaxKind.AtToken, scanner.getStartPos());
+                        atToken.end = scanner.getTextPos();
+                        nextJSDocToken();
 
-                    const tagName = parseJSDocIdentifierName();
-                    if (!tagName) {
+                        const tagName = parseJSDocIdentifierName();
+                        if (!tagName) {
+                            return false;
+                        }
+
+                        switch (tagName.text) {
+                            case "param":
+                                (parentTag.parameterTags || (parentTag.parameterTags = <NodeArray<JSDocParameterTag>>[])).push(handleParamTag(atToken, tagName));
+                                return true;
+                        }
                         return false;
                     }
+                }
 
-                    switch (tagName.text) {
-                        case "type":
-                            if (parentTag.jsDocTypeTag) {
-                                // already has a @type tag, terminate the parent tag now.
-                                return false;
-                            }
-                            parentTag.jsDocTypeTag = handleTypeTag(atToken, tagName);
-                            return true;
-                        case "prop":
-                        case "property":
-                            if (!parentTag.jsDocPropertyTags) {
-                                parentTag.jsDocPropertyTags = <NodeArray<JSDocPropertyTag>>[];
-                            }
-                            const propertyTag = handlePropertyTag(atToken, tagName);
-                            parentTag.jsDocPropertyTags.push(propertyTag);
-                            return true;
+                function scanChildTags<T extends Node>(tryParseChildTag: (parentTag: T) => boolean, parentTag: T): T {
+                    let resumePos = scanner.getStartPos();
+                    let canParseTag = true;
+                    let seenAsterisk = false;
+                    let parentTagTerminated = false;
+
+                    while (token !== SyntaxKind.EndOfFileToken && !parentTagTerminated) {
+                        nextJSDocToken();
+                        switch (token) {
+                            case SyntaxKind.AtToken:
+                                if (canParseTag) {
+                                    parentTagTerminated = !tryParseChildTag(parentTag);
+                                    if (!parentTagTerminated) {

```

Comment:
can you invert `parentTagTerminated` and rename it? I'd much rather read `parentTagContinues = tryParseChild(parentTag); if (parentTagContinues) resumePos = scanner.getStartPos();`